### PR TITLE
kserve migration script #15

### PR DESCRIPTION
**Pull request by Suresh-Nakkeran**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #47 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: hack/kserve_migration.sh [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+if [ ! -z "$knative_services" ]; then
+    mkdir -p ${KSVC_CONFIG_DIR}
+    IFS=','; ksvc_split=(${knative_services}); unset IFS;
+    ksvc_ns=(${ksvc_split[0]})
+    ksvc_names=(${ksvc_split[1]})
+fi
+ksvc_count=${#ksvc_names[@]}
+
+if [ $isvc_count != $ksvc_count ]; then
+    echo "error: inference and knative services counts should be equal."
```
Comment (yuzisun) [inline hack/kserve_migration.sh:73]: This is not sure as for the inference service with transformer or explainer case, there will two underlying knative services for a given inference service

File: hack/kserve_migration.sh [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+# Extract inference service uids
+echo "extracting inference service uids"
+declare -A infr_uid_map
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    infr_uid_map[${isvc_names[$i]}]=$(kubectl get isvc ${isvc_names[$i]} -n ${isvc_ns[$i]} -o json | jq --raw-output '.metadata.uid')
+done
+
+# Update knative services with new owner reference
+echo "updating knative services with new owner reference"
```
Comment (yuzisun) [inline hack/kserve_migration.sh:123]: we also need to update the owner reference for istio virtual service which ones are owned by inference service

File: hack/kserve_migration.sh [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+if [ $isvc_count != $ksvc_count ]; then
+    echo "error: inference and knative services counts should be equal."
+    exit 1;
+fi
+
+# Deploy kserve
+echo "deploying kserve"
+cd ..
+KSERVE_CONFIG=kfserving.yaml
+if [ ${KSERVE_VERSION:3:1} -gt 6 ]; then KSERVE_CONFIG=kserve.yaml; fi
```
Comment (yuzisun) [inline hack/kserve_migration.sh:81]: we do not need the check as we always want to install kserve here.

File: hack/kserve_migration.sh [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+do
+    ksvc_isvc_map[${ksvc_names[$i]}]=$(kubectl get ksvc ${ksvc_names[$i]} -n ${ksvc_ns[$i]} -o json | jq --raw-output '.metadata.ownerReferences[0].name')
+    kubectl patch ksvc ${ksvc_names[$i]} -n ${ksvc_ns[$i]} --type json -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
+done
+
+# Remove owner references from virtual services
+echo "removing owner references from virtual services"
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    kubectl patch virtualservices ${isvc_names[$i]} -n ${isvc_ns[$i]} --type json -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
```
Comment (yuzisun) [inline hack/kserve_migration.sh:94]: We should check the owner reference's API group is from serving.kubeflow.org, there are other virtual services owned by knative service which we should not change the owner reference

File: hack/kserve_migration.sh [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+done
+
+# Update apiversion in knative services
+echo "updating apiversion in knative services"
+for (( i=0; i<${ksvc_count}; i++ ));
+do
+    kubectl get ksvc ${ksvc_names[$i]} -n ${ksvc_ns[$i]} -o yaml > "${KSVC_CONFIG_DIR}/${ksvc_names[$i]}.yaml"
+    sed -i -- 's/kubeflow.org/kserve.io/g' "${KSVC_CONFIG_DIR}/${ksvc_names[$i]}.yaml"
+    kubectl apply -f "${KSVC_CONFIG_DIR}/${ksvc_names[$i]}.yaml"
+done
```
Comment (yuzisun) [inline hack/kserve_migration.sh:154]: I thought at this point the API version is already updated to kserve.io, do we still need to run this?

File: hack/kserve_migration.sh [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+# Clean up kfserving
+echo "deleting kfserving namespace"
+kubectl delete ns kfserving-system
```
Comment (yuzisun) [inline hack/kserve_migration.sh:158]: I think there are still some non namespace scoped resource which needs to be deleted like cluster rbac and bindings, as well as the webhooks.

Comment (Suresh-Nakkeran) [reply hack/kserve_migration.sh:170]: Added steps to delete non namespace resources.

File: hack/kserve_migration.sh (hunk 5 above)
Comment (Suresh-Nakkeran) [reply hack/kserve_migration.sh]: Yes @yuzisun we can remove this step.

File: hack/kserve_migration.sh (hunk 4 above)
Comment (Suresh-Nakkeran) [reply hack/kserve_migration.sh]: I added the check.

File: hack/kserve_migration.sh [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+cd hack
+echo "kserve deployment completed"
+
+# Remove owner references from knative services
+echo "removing owner references from knative services"
+declare -A ksvc_isvc_map
+for (( i=0; i<${ksvc_count}; i++ ));
+do
+    ksvc_isvc_map[${ksvc_names[$i]}]=$(kubectl get ksvc ${ksvc_names[$i]} -n ${ksvc_ns[$i]} -o json | jq --raw-output '.metadata.ownerReferences[0].name')
+    kubectl patch ksvc ${ksvc_names[$i]} -n ${ksvc_ns[$i]} --type json -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
```
[the remark this answers is not in the document]
Comment (Suresh-Nakkeran) [reply hack/kserve_migration.sh]: I added the check.

File: hack/kserve_migration.sh [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+echo "deploying inference services on kserve"
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.annotations'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.creationTimestamp'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.finalizers'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.generation'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.resourceVersion'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.uid'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'status'
```
[the remark this answers is not in the document]
Comment (Suresh-Nakkeran) [reply hack/kserve_migration.sh:114]: @yuzisun I am using Kubernete v1.20.2. I do not see managedFields in metadata in inferenceservice yaml file.

File: hack/kserve_migration.sh [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        fi
+    done
+}
+
+# Checks whether the kfserving is running or not
+isControllerRunning kfserving-system
+
+# Get inference services config
+echo "getting inference services config"
+inference_services=$(kubectl get isvc -A -o jsonpath='{.items[*].metadata.namespace},{.items[*].metadata.name}')
```
Comment (yuzisun) [inline hack/kserve_migration.sh:43]: I got an issue when both kfserving/kserve are installed, isvc becomes ambiguous and you need to qualify the resource with `inferenceservices.serving.kubeflow.org`

File: hack/kserve_migration.sh [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+KSERVE_CONFIG=kserve.yaml
+for i in 1 2 3 4 5 ; do kubectl apply -f install/${KSERVE_VERSION}/${KSERVE_CONFIG} && break || sleep 15; done
+kubectl wait --for=condition=ready --timeout=120s po --all -n kserve
+isControllerRunning kserve
+cd hack
+echo "kserve deployment completed"
+
+# Remove owner references from knative services
+echo "removing owner references from knative services"
+declare -A ksvc_isvc_map
```
[the remark this answers is not in the document]
Comment (Suresh-Nakkeran) [reply hack/kserve_migration.sh]: @yuzisun declare: -A requires bash version >=4. I created an image and added dependencies.

File: hack/kserve_migration.sh (hunk 9 above)
Comment (Suresh-Nakkeran) [reply hack/kserve_migration.sh]: @yuzisun updated.

File: hack/kserve-migration/kserve_migration.sh [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.creationTimestamp'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.finalizers'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.generation'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.resourceVersion'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.uid'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.managedFields'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'status'
+    sed -i -- 's/kubeflow.org/kserve.io/g' ${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml
+    kubectl apply -f "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml"
+done
```
Comment (yuzisun) [inline hack/kserve-migration/kserve_migration.sh:123]: Move this section before removing the owner reference

File: hack/kserve-migration/kserve_migration.sh [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        kubectl patch virtualservices ${isvc_names[$i]} -n ${isvc_ns[$i]} --type json -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
+    fi
+done
+sleep 5
+
+# Deploy inference services on kserve
+echo "deploying inference services on kserve"
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.annotations'
```
Comment (yuzisun) [inline hack/kserve-migration/kserve_migration.sh:113]: do not remove the entire annotations as some annotation is used by isvc, you will need to remove the `lastAppliedAnnotation`

File: hack/kserve-migration/kserve_migration.sh [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.creationTimestamp'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.finalizers'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.generation'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.resourceVersion'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.uid'
+    yq d -i "${ISVC_CONFIG_DIR}/${isvc_names[$i]}.yaml" 'metadata.managedFields'
```
Comment (yuzisun) [inline hack/kserve-migration/kserve_migration.sh:119]: `yq` does not exit out if certain fields like `managedFields` does not exist for earlier kubernetes release right?

Comment (Suresh-Nakkeran) [reply hack/kserve-migration/kserve_migration.sh:119]: yes @yuzisun

File: hack/kserve-migration/kserve_migration.sh (hunk 12 above)
Comment (Suresh-Nakkeran) [reply hack/kserve-migration/kserve_migration.sh]: @yuzisun I updated.

**Comment by yuzisun** [comment]
moved to kserve repo

**Pull request closed without merging**